Unified attention CK Tile kernel (#3128)
ROCm/composable_kernel: example/ck_tile/42_unified_attention/README.md

# fused multi-head attention

This folder contains examples for unified attention (fused multi-head attention) using the ck_tile tile-programming implementation. The examples demonstrate the usage of the tile-programming API, as well as the new approach to constructing kernel templates and instantiating them. 

## build

> ```

Reviewer: ```suggestion
## build

```
```

> This folder contains example for unified attention (fused multi-head attention) using ck_tile tile-programming implementation. It is a good example to demonstrate the usage of tile-programming API, as well as illustrate the new approach to construct a kernel template and instantiate it(them) while keeping compile time fast.

Reviewer 2: ```suggestion
This folder contains examples for unified attention (fused multi-head attention) using the ck_tile tile-programming implementation. The examples demonstrate the usage of the tile-programming API, as well as the new approach to constructing kernel templates and instantiating them. 
```

Wasn't sure if the compile times were fast because of the new way that kernels were instantiated, or if it demonstrated a new way to keep compile times fast independent of instantiation.

> ## build

Reviewer 2: Add an empty line below this

# in the root of ck_tile
mkdir build && cd build
# you can replace <arch> with the appropriate architecture (for example gfx90a or gfx942) or leave it blank
../script/cmake-ck-dev.sh  ../ <arch>
make tile_example_unified_attention -j

> ```

Reviewer 2: Add an empty line below this

> ```
> This will result in an executable `build/bin/tile_example_unified_attention`

Reviewer: ```suggestion
```

This will result in an executable `build/bin/tile_example_unified_attention`
```

> ## kernel

Reviewer 2: Put an empty line beneath all headers

> 
> ## kernel
> The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.

Reviewer: ```suggestion
## kernel

The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.
```

> The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.

Reviewer 2: This is really confusing.

Do you mean the original version of CK or some pre-release version of CK Tile?

> There are 2 template parameters for this kernel template.

Reviewer 2: Add an empty line below

> 
> There are 2 template parameters for this kernel template.
> * `FmhaPipeline` is one of the block_tile_pipeline(under `include/ck_tile/tile_program/block_tile_pipeline`) which is a performance critical component. Indeed, we did a lot of optimization and trials to optimize the pipeline and may still workout more performance pipeline and update into that folder. People only need to replace this pipeline type and would be able to enjoy the benefit of different performant implementations (stay tuned for updated pipeline(s)).

Reviewer: ```suggestion
There are 2 template parameters for this kernel template.

* `FmhaPipeline` is one of the block_tile_pipeline(under `include/ck_tile/tile_program/block_tile_pipeline`) which is a performance critical component. Indeed, we did a lot of optimization and trials to optimize the pipeline and may still workout more performance pipeline and update into that folder. People only need to replace this pipeline type and would be able to enjoy the benefit of different performant implementations (stay tuned for updated pipeline(s)).
```

> * `FmhaPipeline` is one of the block_tile_pipeline(under `include/ck_tile/tile_program/block_tile_pipeline`) which is a performance critical component. Indeed, we did a lot of optimization and trials to optimize the pipeline and may still workout more performance pipeline and update into that folder. People only need to replace this pipeline type and would be able to enjoy the benefit of different performant implementations (stay tuned for updated pipeline(s)).

Reviewer 2: This is a really confusing paragraph. 

Is `FmhaPipeline` an enum of `block_tile_pipeline` class? If so, then it should be referred to as such. If not, can you explain what it is because it's not clear.

The rest reads like an internal note. I don't understand what it's trying to say. Can you clarify?

> * `EpiloguePipeline` will modify and store out the result in the last phase. People usually will do lot of post-fusion at this stage, so we also abstract this concept. Currently we didn't do much thing at the epilogue stage but leave the room for future possible support.

Reviewer 2: ```suggestion
* `EpiloguePipeline` is the last stage of the pipeline. It modifies and stores the result. Post-fusion can be done at this stage though the example only returns the result.
```

> ## codegen

Reviewer 2: new line

> To speed up compile time, we instantiate the kernels into separate file. In this way we can benefit from parallel building from CMake/Make system. This is achieved by `generate.py` script. Besides, you can look into this script to learn how to instantiate a kernel instance step by step, which is described in `FMHA_FWD_KERNEL_BODY` variable.

Reviewer 2: ```suggestion
To speed up compile time, the `generate.py` script splits the kernel into separate files. This allows the example to take advantage of CMake/Make parallel building.  
```

What does `FMHA_FWD_KERNEL_BODY` describe exactly?

> ## executable

Reviewer 2: newline

> `tile_example_unified_attention` is the example executable, implemented in `fmha_fwd.cpp`. You can type `./bin/tile_example_unified_attention -?` to list all the arguments. Below is an example of the output (may subject to change)

Reviewer 2: ```suggestion
Building `fmha_fwd.cpp` yields the `tile_example_unified_attention` executable. Use `./bin/tile_example_unified_attention -?` to list all the arguments. 
```

> ```

Reviewer 2: newline above

args:

>           -v    weather do CPU validation or not (default:1)

Reviewer 2: ```suggestion
          -v    Perform CPU validation (default:1)
```

What are the possible values? Is the default to perform validation or not perform validation?

       -mode    kernel mode. 0:batch, 1:group (default:0)
          -b    batch size (default:2)

>           -h    num of head, for q (default:8)

Reviewer 2: Is this number of heads? is there a lower or upper limit?

        -h_k    num of head, for k/v, -1 means equal to h (default:-1)

>                 if not equal to h, then this is GQA/MQA case

Reviewer 2: ```suggestion
        -h_k    Number of heads for k/v. When set to -1, the number of heads will be set to the same value as was set for q. (default:-1)
```

What is the GQA/MQA case?

          -s    seqlen_q. if group-mode, means the average value of seqlen_q (default:3328)
                total_seqlen_q = seqlen_q * batch, and seqlen_q per batch may vary

>                 also with "-s=s0,s1,s2..." comma seperated int to set per batch seqlen(group-mode)

Reviewer 2: ```suggestion
  seqlen_q. When in group-mode, this is the average value of seqlen_q. In group mode it can also be set to a series of comma-separated integers to set seqlen per batch.  (default:3328)```

Was the equation supposed to be total? The description mentions an average not a total:

total_seqlen_q = seqlen_q * batch 

Also, what can this be set to when not in group mode? Or is it not set at all in kernel mode?

Given that this can also be set to a series of comma-separated ints, why is the default such a large number (3328). Does the number have meaning?

        -s_k    seqlen_k (including new key/value), -1 means equal to s (default:-1)

>                 also with "-s_k=s0,s1,s2..." comma-separated ints to set seqlen per batch (group mode)

Reviewer 2: ```suggestion
-s_k    seqlen_k. Set this to a list of comma-separated integers for seqlen per bathch, or set to -1 to assign the same values as seqlen_q (default:-1)
```

     -s_qpad    seqlen_q stride between 2 batches (group-mode optional) (default:-1)

>                 Provide positive strides per-batch to simulate physical padding on Q

Reviewer 2: ```suggestion
seqlen_q stride size between 2 batches. Simulates physical padding on Q. Default -1.
```

What does a negative value, in particular the default value, do?

>      -s_kpad    seqlen_k stride between 2 batches, currently used in group-mode only  (default:-1)

Reviewer 2: need to say what -1 does

                for kv-cache case, each batch [1,s,h,d]/[1,h,s,d] can have a stride
                along seqlen, instead of packed, same as xformer kv_padding,
                must be greater than or equal to s_k

>           -d    head dim for q, k (default:128)

Reviewer 2: ```suggestion
          -d    head dimension for q, k (default:128)
```

Need to say if there's a lower or upper bound

>         -d_v    head dim for v, -1 means equal to d (default:-1)

Reviewer 2: need to say whether there's a lower or upper bound

    -scale_s    scale factor of S. 0 means equal to 1/sqrt(hdim). (default:0)
                note when squant=1, this value will be modified by range_q/k
    -range_q    per-tensor quantization range of q. used if squant=1. (default:16)
    -range_k    per-tensor quantization range of k. used if squant=1. (default:16)
    -range_v    per-tensor quantization range of v. used if squant=1. (default:16)
    -range_p    per-tensor quantization range of p [e^(s-m)]. used if squant=1. (default:1)

>     -range_o    per-tensor quantization range of o (p*v). used if squant=1. (default:16)

Reviewer 2: in general, you should state if there are lower or upper bounds on the values

     -squant    if using static quantization fusion or not. auto: fp8 will default use squant, other will not (default:auto)
                0: no static quant(not implemented) 1: apply scale_p and scale_o with respect to P and O.

>                 calculate scale_s, scale_p, scale_o according to range_q, range_k, range_v, range_p, range_o

Reviewer 2: ```suggestion
When set to 0, no static quantization fusion is applies. When set to 1, scale_p and scale_o are applied with respect to P and O. When set to auto, only fp8 will apply static quantization. (default:auto)
```

Not sure how the last line about calculating those values fit in.

      -iperm    permute input (default:1)

>                 if true, will be b*h*s*d, else b*s*h*d

Reviewer 2: ```suggestion
When set to 1, input will be permuted with b*h*s*d. When set to 0, will be b*s*h*d. Default 1.
```

I'm assuming 1==true and 0==false

>       -operm    permute output (default:1)

Reviewer 2: need the values and the results

       -bias    n or 0, no bias (default:n)
                e(lementwise) or 1, elementwise bias with 1*1*s*s. e:1, 1*h*s*s. e:2, b*h*s*s

>                 a(libi) or 2, alibi with 1*h. a:1, b*h

Reviewer 2: ```suggestion
Set to  n or 0 for no bias. Set to e or 1 for elementwise bias with 1*1*s*s. Set to e:1 for elementwise bias with 1*h*s*s. Set to e:2 for elementwise bias with b*h*s*s. Set to a or 2 for alibi bias with 1*h. Set to a:1 for alibi bias with b*h. Default is 0, no bias.```

>        -prec    data type. fp16/bf16/fp8/bf8 (default:fp16)

Reviewer 2: what type of data type? for what? Needs clarification

       -mask    0: no mask, 1: top-left(same as 't'), 2:bottom-right(same as 'b') (default:0)
                't', top-left causal mask, 'b', bottom-r causal mask
                't:l,r', top-left sliding window attn(swa) with FA style left right size
                'b:l,r', bottom-r sliding window attn(swa) with FA style left right size
                'xt:window_size', xformer style masking from top-left, window_size negative is causal, positive is swa
                'xb:window_size', xformer style masking from bottom-r, window_size negative is causal, positive is swa

>                 'g:y,x', generic attention mask coordinate with y/x size (only debug purpose for now)

Reviewer 2: ```suggestion
 Set to 0 for no mask. Set to 1 top-left mask. Set to 2 for bottom-right mask. Set to 't' for top-left causal mask. Set to 'b' for bottom-right causal mask. Set to 't:l,r' for top-left sliding window attn(swa) with FA style left right size mask. Set to 'b:l,r' for bottom-r sliding window attn(swa) with FA style left right size mask. Set to 'xt:window_size' for an xformer style mask from top-left where setting window_size to a negative value results in a causal mask and setting it to a positive value with result in an swa mask. Set to 'xb:window_size' for an xformer style mask from bottom-right where setting window_size to a negative value results in a causal mask and setting it to a positive value with result in an swa mask. Set to  'g:y,x' for generic attention mask coordinates with y/x size (only for debugging purposes). Default is 0, no mask.
```

>     -vlayout    r for row-major(seqlen*hdim), c for col-major(hdim*seqlen) (default:r)

Reviewer 2: ```suggestion
    -vlayout    Set to r for row-major. Set to c for column-major. (default:r)
```

>         -lse    0 not store lse, 1 store lse (default:0)

Reviewer 2: ```suggestion
        -lse    Set to 0 to not store lse. Set to 1 to store lse (default:0)
```

      -kname    if set to 1 will print kernel name (default:0)
       -init    init method. ui, uniform random int, ni, normalized random int (default:uf)

>                 uf, uniform random float, nf, normalized random float, tf, trig float, uf:q, quantization

Reviewer 2: ```suggestion
 init method. Set to ui for a uniform random int. Set to ni for a normalized random int. Set to uf for a uniform random float. Set to nf for a normalized random float. Set to  tf for a trig float. Set to uf:q for quantization. Default uf.
```

>        -seed    random seed used for initializing input tensors. 0 for non-deterministic seed (default:11939)

Reviewer 2: Is there a reason for that default?

  -drop_seed    seed for random number generator (default:1)
-drop_offset    offset for random number generator (default:0)

>  -drop_prefs    seed and offset values are present on GPU; 0 - host, 1 - device/GPU (default:0)

Reviewer 2: I don't quite understand what this option is. Can you clarify?

>  -num_splits    number of splits for key/value. 0 to determine actual number by heuristic (default:1)

Reviewer 2: What is 1? What does the default do?

>      -warmup    number of iterations before benchmark the kernel (default:5)

Reviewer 2: ```suggestion
     -warmup    number of iterations before benchmarking the kernel (default:5)
```

>      -repeat    number of iterations to benchmark the kernel (default:20)

Reviewer 2: ```suggestion
     -repeat    number of times to iterate when benchmarking the kernel (default:20)
```

For this and warmup, any upper or lower bounds should be mentioned.

>        -json    0: No Json, 1: Dump Results in Json format (default:0)

Reviewer 2: ```suggestion
       -json    Set to 1 to write out results in to a json file. (default:0, results are not written to a json file)
```

Where do the results go when set to 0

>    -jsonfile    json file name to dump results (default:fmha_fwd.json)

Reviewer 2: ```suggestion
   -jsonfile    json file to which to write results when `-json` is set to 1 (default:fmha_fwd.json)
```

 -q_eff_lens    Batch-mode only: per-batch effective seqlen for Q (exclude PAD) (default:"")

>                 Comma-separated list of length 'b'. If empty, no override

Reviewer 2: I don't quite understand what this is. What is b in this case?

-kv_eff_lens    Batch-mode only: per-batch effective seqlen for KV (exclude PAD) (default:"")

>                 Comma-separated list of length 'b'. If empty, no override

Reviewer 2: See previous comment.

> ```

Reviewer 2: insert newline

Example 1: `./bin/tile_example_unified_attention -b=1 -h=16 -s=16384 -d=128` will run a fmha case with batch=1, nhead=16, sequence length=16384, hdim=128, fp16 case.
Example 2: `./bin/tile_example_unified_attention -b=1 -h=8 -s=16384 -d=64 -drop_prefs=1 -drop_seed=10 -drop_offset=1234` will run a fmha case with 
  batch=1, nhead=8, sequence length=16384, hdim=64, drop_seed=0 (in GPU memory), drop_offset=1234 (in GPU memory) fp16 case

> ## Padding Examples

Reviewer 2: newline

Example 3 (Group mode with padding): `./bin/tile_example_unified_attention -mode=1 -b=2 -h=8 -s=1024,2048 -s_k=1024,2048 -s_qpad=1536,3072 -s_kpad=1536,3072 -d=128` will run group mode with 2 batches having different sequence lengths (1024, 2048) but physically padded to (1536, 3072) respectively.

Example 4 (Batch mode with effective lengths): `./bin/tile_example_unified_attention -mode=0 -b=2 -h=8 -s=2048 -s_k=2048 -d=128 -q_eff_lens=1024,1536 -kv_eff_lens=1024,1536` will run batch mode where all batches use 2048 as physical sequence length but have effective lengths of (1024, 1536) for Q and KV respectively.

> ## support features

Reviewer 2: newline

> Currently we are still in rapid development stage, so more features/optimizations will be coming soon.

Reviewer 2: ```suggestion
This is still in the development phase and new features and optimizations are being added.
```

> ### hdim

Reviewer 2: newline

> Currently we support `32/64/128/256` hdim for `fp16`/`bf16`, within which `64`/`128` is better optimized. hdim should be multiple of 8, while seqlen_s can be arbitrary. For hdim be arbitrary number, it can be support through padding kernel of `qr` pipeline (we didn't generate this in generate.py by default)

Reviewer 2: ```suggestion
Currently we support `32/64/128/256` hdim for `fp16`/`bf16`, optimized for `64`/`128`. hdim should be set to a multiple of 8, while seqlen_s can be set to any value. 
```

Not sure what this means:

"For hdim be arbitrary number, it can be support through padding kernel of `qr` pipeline (we didn't generate this in generate.py by default)"

> ### group/batch mode

Reviewer 2: newline

> Currently we support both `batch mode` and `group mode` (or `varlen`, in FA's term), by setting `-mode` = `0` or `1`. In `group mode` different kind of attention mask is also supported(see below)

Reviewer 2: ```suggestion
Currently we support both batch mode and group mode (or varlen, in FA's term). Different types of attention masking can be applied group mode.
```

> ### MQA/GQA

Reviewer 2: newline

> By setting `-h`(nhead for q) and `-h_k`(nhead for k/v) with different number, you can achieve MQA/GQA. Please pay attention that `h % h_K == 0` when you set different numbers.

Reviewer 2: This is very vague. Can it be clarified?

> ### input/output permute, and `b*s*3*h*d`

Reviewer 2: newline

> If you look at the kernel argument inside `fmha_fwd_kernel.hpp`, we support providing arbitrary stride for seqlen(stride_q/k/v), nhead, batch of q/k/v matrix, hence it is very flexible to support `b*h*s*d` or `b*s*h*d` input/output permute. The `-iperm=0/1`, `-operm=0/1` is a convenient way to achieve this through the executable. We didn't provide a command-line arg to test `b*s*3*h*d` layout which is by default used by torch/FA, but it's trivial to achieve this if one set the proper `stride_q/k/v` value as `3*h*d`.

Reviewer 2: ```suggestion
The kernel argument inside `fmha_fwd_kernel.hpp` supports arbitrary strides for seqlen (stride_q/k/v), nhead, batch of q/k/v matrix, making it possible to support `b*h*s*d` or `b*s*h*d` input/output permutations.  The `-iperm=0/1`, `-operm=0/1` options is a convenient way to achieve this. There is no option to test `b*s*3*h*d` layout which is by default used by torch/FA, but it's possible to do this by setting the proper `stride_q/k/v` value as `3*h*d`.
```

Does this have to be set in the code itself? 

also: in general it's not a good idea to say something is trivial. Not all users might find something trivial.

> ### attention bias

Reviewer 2: newline

> Attention bias is supported with the layout of `1*1*s*s`(similiar to input/output, different layout can be supported by changing the stride value for bias, or even extend to `b*h*s*s`) and bias value in float number.

Reviewer 2: This is very confusing. Can this be clarified?

> ### alibi

Reviewer 2: newline

alibi is supported

> ### lse

Reviewer 2: newline

> For training kernels, "log sum exp" need to store out in forward and used in backward. We support this by setting `-lse=1`

Reviewer 2: This is confusing. Can it be clarified a bit?

> ### vlayout

Reviewer 2: newline

> We support v matrix in both row-major(`seqlen*hdim`) and col-major(`hdim*seqlen`). Since the accumulate(reduce) dimension for V is along `seqlen`, for current AMD's mfma layout which expect each thread to have contiguous register holding pixels along reduce dimension, it's easier to support col-major V layout. However, the performance of col-major is not necessarily faster than row-major, there are many factors that may affect the overall performance. We still provide the `-vlayout=r/c` here to switch/test between different layouts.

Reviewer 2: ```suggestion
We support V matrix in both row-major and column-major. Since the accumulate (reduce) dimension for V is along `seqlen`, and AMD's mfma layout expects each thread to have contiguous registers holding pixels along the reduce dimension, it's easier to support col-major V layout. However, the performance of col-major is not necessarily faster than row-major. There are many factors that may affect the overall performance. We still provide the `-vlayout=r/c` here to switch/test between different layouts.
```

When you say pixels, do you mean data?

> ### attention mask

Reviewer 2: newline

> we support `causal mask` and `sliding window attention(swa)` mask in both batch and group mode, either from top-left or bottom-right.

Reviewer 2: ```suggestion
Causal and sliding window attention(swa) masks in both batch and group mode are supported, both from either top-left or bottom-right.
```

> Underneath, we unify the mask expression into `generic attention mask coordinate`, providing an uniformed approach for each batch to locate the corresponding pixel need to be masked out.

Reviewer 2: ```suggestion
The mask expression is unified into generic attention mask coordinates, providing an uniform approach for each batch to locate the corresponding pixel need to be masked out.
```

Is this pixels? Is this example for image masking? If that's the case, it should be made clear earlier on.

![](misc/gamc.png)

> Since FA/xformer style with window_size_left/right is more popular, we accept window_size as parameter and convert that internally to our generic coordinate(this coordinate can express more cases). Below shows some example of how to achieve different kind of mask through cmdline.

Reviewer 2: ```suggestion
Since FA/xformer style with window_size_left/right is more popular, we accept window_size as parameter and convert that internally to our generic coordinate. The following are some examples of how to achieve different kind of mask through cmdline:
```

| mask case|  cmdline    | FA style | xformer style |
|----------|:-------------:|:-------------:|:-------------:|
| no mask |  `-mask=0`(default) | | |
| causal mask from top-left | `-mask=1` or `-mask=t` | `-mask=t:-1,0` | `-mask=xt:-1` |
| causal mask from bottom-right | `-mask=2` or `-mask=b` | `-mask=b:-1,0` | `-mask=xb:-1` |
| swa from top-left | | `-mask=t:3,5` | `-mask=xt:4` |
| swa from bottom-right | |  `-mask=b:10,11` | `-mask=xb:16` |

Note FA use bottom-right by default to express swa case, here we require you explicitly specify top-left/bottom-right.

> ### dropout

Reviewer 2: newline

TBD

> ### sequence padding and variable length support

Reviewer 2: newline

We support sequence padding and variable-length processing in both batch and group modes fmha forward to handle real-world scenarios where sequences have different lengths.

> **Group Mode Padding**: Use `-s_qpad` and `-s_kpad` to specify physical stride between batches, enabling padded layouts. Each batch can have different logical sequence lengths (`-s`, `-s_k`) but use larger physical strides for memory alignment.

Reviewer 2: ```suggestion
**Group Mode Padding**: Use `-s_qpad` and `-s_kpad` to specify physical strides between batches and enable padded layouts. Each batch can have different logical sequence lengths (`-s`, `-s_k`). Use larger physical strides for memory alignment.
```

**Batch Mode Variable Length**: Use `-q_eff_lens` and `-kv_eff_lens` to specify effective sequence lengths per batch. All batches share the same physical sequence length, but the kernel processes only the effective portions. This enables efficient variable-length attention without memory waste.

Both approaches optimize memory access patterns while supporting flexible sequence length requirements commonly found in transformer inference scenarios.

> ## FP8 experimental support

Reviewer 2: newline

    -vlayout    r for row-major(seqlen*hdim), c for col-major(hdim*seqlen) (default:r)
        -lse    0 not store lse, 1 store lse (default:0)
      -kname    if set to 1 will print kernel name (default:0)
       -init    init method. ui, uniform random int, ni, normalized random int (default:uf)
                uf, uniform random float, nf, normalized random float, tf, trig float, uf:q, quantization
       -seed    random seed used for initializing input tensors. 0 for non-deterministic seed (default:11939)
  -drop_seed    seed for random number generator (default:1)
-drop_offset    offset for random number generator (default:0)
 -drop_prefs    seed and offset values are present on GPU; 0 - host, 1 - device/GPU (default:0)
 -num_splits    number of splits for key/value. 0 to determine actual number by heuristic (default:1)
     -warmup    number of iterations before benchmark the kernel (default:5)
     -repeat    number of iterations to benchmark the kernel (default:20)
       -json    0: No Json, 1: Dump Results in Json format (default:0)
   -jsonfile    json file name to dump results (default:fmha_fwd.json)

>  -q_eff_lens    Batch-mode only: per-batch effective seqlen for Q (exclude PAD) (default:"")

Reviewer 3: Please remove `q_eff_lens`/`kv_eff_lens` from the README.md if they are not intended to be supported.
